reader_concurrency_semaphore.hh

```diff
@@ -151,28 +152,33 @@ private:
         resources res;
         entry(promise<lw_shared_ptr<reader_permit>>&& pr, resources r) : pr(std::move(pr)), res(r) {}
     };
-    struct expiry_handler {
+
+    class expiry_handler {
+        sstring _semaphore_name;
+    public:
+        explicit expiry_handler(sstring semaphore_name)
+            : _semaphore_name(std::move(semaphore_name)) {}
         void operator()(entry& e) noexcept {
-            e.pr.set_exception(semaphore_timed_out());
+            e.pr.set_exception(named_semaphore_timed_out(_semaphore_name));
         }
     };
 
+    future<lw_shared_ptr<reader_concurrency_semaphore::reader_permit>>
+    make_queue_overflow_exception_future() const;
+
 private:
     resources _resources;
 
     expiring_fifo<entry, expiry_handler, db::timeout_clock> _wait_list;
 
+    sstring _name;
     size_t _max_queue_length = std::numeric_limits<size_t>::max();
-    std::function<std::exception_ptr()> _make_queue_overloaded_exception;
+    std::function<void()> _prethrow_action = nullptr;
```

Comment:
`std::function<>` has a default constructor, so this initialization is not needed.

Reply:
Removed (although I did it on purpose)